### DISCUSSION TITLE
How would I add something like a "--version" argument to a program?

### DISCUSSION THREAD

**Discussion by nessus42** [bkirwi/decline#339] [Q&A] [answered]
I'm writing a CLI program that has a complex bunch of arguments, many of which are required. But kind of like how "--help" bypasses the necessity for the rest of the command line to exist, I'd also like a "--version" option that would behave similarly and just cause the program to output its version number and exit successfully. How can I accomplish this in any easy manner?

I suppose that I could always include the version in the program summary that is output by "--help", but ideally I'd like to have a separate option for the version number.

**Comment by bkirwi** [comment, marked as answer]
So, two answers:

- If you're using `CommandApp`, you can pass a `version` string to the constructor, like so: https://github.com/bkirwi/decline/blob/f6752ea7ed67413c426cd7022cb72eeecb6ba0d5/core/jvm/src/test/scala/com/monovore/example/decline/ListDir.scala#L7-L10
- You can implement your own `--version` option quite easily also:

```scala
let programOpts = ??? // your normal program options go here
let versionOpts = Opts.flag("version", "Print version information and exit").map(_ => println("Version 1.0"))
let opts = programOpts orElse versionOpts
```

`CommandApp`'s version flag is actually implemented like this under the hood.

Many people start by just using `orElse` for subcommands, but it's a pretty powerful feature. I endorse reaching for it anytime you have mutually-exclusive sets of options.

**Reply by nessus42** [reply]
Thanks!!!
